management/api-client.go

```diff
@@ -0,0 +1,109 @@
+package ziti
+
+import (
+	"crypto/tls"
+	"github.com/go-openapi/runtime"
+	openapiclient "github.com/go-openapi/runtime/client"
+	"github.com/go-openapi/strfmt"
+	"github.com/openziti/edge-api/rest_management_api_client"
+	"github.com/openziti/edge-api/rest_util"
+	"net"
+	"net/http"
+	"net/http/cookiejar"
+	"net/url"
+	"time"
+)
+
+type AuthenticatedApiClient struct {
+	*ApiClient
+	ZtSession     string
+	Authorization []string
+}
+
+func (a AuthenticatedApiClient) AuthenticateRequest(request runtime.ClientRequest, _ strfmt.Registry) error {
+	if err := request.SetHeaderParam("zt-session", a.ZtSession); err != nil {
+		return err
+	}
+
+	if err := request.SetHeaderParam("Authorization", a.Authorization...); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func NewAuthenticatedApiClient(apiSessionToken string, apiClient *ApiClient) *AuthenticatedApiClient {
+	authenticatedClient := &AuthenticatedApiClient{
+		ApiClient: apiClient,
+		ZtSession: apiSessionToken,
+	}
+
+	apiClient.Runtime.DefaultAuthentication = authenticatedClient
+
+	return authenticatedClient
+}
+
+type ApiClient struct {
+	*rest_management_api_client.ZitiEdgeManagement
+
+	Runtime       *openapiclient.Runtime
+	HttpClient    *http.Client
+	HttpTransport *http.Transport
+	TlsConfig     *tls.Config //use SetTlsConfig() to replace
+}
+
+func NewApiClient(clientApiUrl *url.URL) *ApiClient {
+	httpClient, httpTransport, tlsConfig := NewHttpComponents()
+
+	apiClient, apiRuntime := NewApiComponents(httpClient, clientApiUrl)
+
+	return &ApiClient{
+		ZitiEdgeManagement: apiClient,
+		Runtime:            apiRuntime,
+		HttpClient:         httpClient,
+		HttpTransport:      httpTransport,
+		TlsConfig:          tlsConfig,
+	}
+}
+
+func NewApiComponents(httpClient *http.Client, clientApiUrl *url.URL) (*rest_management_api_client.ZitiEdgeManagement, *openapiclient.Runtime) {
+	apiClientRunTime := openapiclient.NewWithClient(clientApiUrl.Host, clientApiUrl.Path, rest_management_api_client.DefaultSchemes, httpClient)
+
+	apiClient := rest_management_api_client.New(apiClientRunTime, nil)
+
+	return apiClient, apiClientRunTime
+}
+
+func NewHttpComponents() (*http.Client, *http.Transport, *tls.Config) {
```

Comment:
i feel like these `NewApiComponents` is growing the api needlessly and adds confusion. would you consider making this private?

Reply:
mmmmm I can do something to reduce the exposed names. A sub-package would do the same thing.